gitlabform/gitlabform/test/test_group_shared_with.py

```diff
@@ -0,0 +1,164 @@
+
+    # we try to remove all users, not just those added above,
+    # on purpose, as more may have been added in the tests
+    for user in users:
+        gitlab.remove_member_from_group(group, user)
+
+
+class TestGroupSharedWith:
+    def test__add_group(self, gitlab, group, users, groups, one_owner):
+        no_of_members_before = len(gitlab.get_group_members(group))
+
+        add_shared_with = f"""
+        projects_and_groups:
+          {group}/*:
+            group_members:
+              {users[0]}:
+                access_level: 50
+            group_shared_with:
+              {groups[0]}:
+                group_access_level: 30
+              {groups[1]}:
+                group_access_level: 30
+        """
+
+        run_gitlabform(add_shared_with, group)
+
+        members = gitlab.get_group_members(group)
+        assert len(members) == no_of_members_before, members
+
+        shared_with = gitlab.get_group_shared_with(group)
+        assert len(shared_with) == 2
+
+    def test__remove_group(self, gitlab, group, users, groups, one_owner):
+
+        no_of_members_before = len(gitlab.get_group_members(group))
+        no_of_shared_with_before = len(gitlab.get_group_shared_with(group))
+
+        remove_users = f"""
+        projects_and_groups:
+          {group}/*:
+            enforce_group_members: true
+            group_members:
+              {users[0]}:
+                access_level: 50
+            group_shared_with:
+              {groups[0]}:
+                group_access_level: 30
+        """
+
+        run_gitlabform(remove_users, group)
+
+        members = gitlab.get_group_members(group)
+        assert len(members) == no_of_members_before
+
+        shared_with = gitlab.get_group_shared_with(group)
+        assert len(shared_with) == no_of_shared_with_before - 1
+
+        assert [sw["group_name"] for sw in shared_with] == [groups[0]]
+
+    def test__not_remove_groups_with_enforce_false(
+        self, gitlab, group, users, groups, one_owner
+    ):
+
+        no_of_members_before = len(gitlab.get_group_members(group))
+        no_of_shared_with_before = len(gitlab.get_group_shared_with(group))
+
+        setups = [
+            # flag explicitly set to false
+            f"""
+            projects_and_groups:
+              {group}/*:
+                enforce_group_members: false
+                group_members:
+                  {users[0]}:
+                    access_level: 50
+                group_shared_with:
+                  {groups[0]}:
+                    group_access_level: 30
+            """,
+            # flag not set at all (but the default is false)
+            f"""
+            projects_and_groups:
+              {group}/*:
+                group_members:
+                  {users[0]}:
+                    access_level: 50
+                group_shared_with:
+                  {groups[0]}:
+                    group_access_level: 30
+            """,
+        ]
```

Comment:
I think that these 2 examples would not remove the sharing even with `enforce_group_members: true`. Isn't the `group_shared_with` content in them exactly the same as at the end of the previous test:
```
            group_shared_with:
              {groups[0]}:
                group_access_level: 30
```
?

Reply:
Thanks, good spot, I've corrected this.